### PR TITLE
[BEAM-588] Implement profiling for Python DirectRunner and SDK Harness.

### PR DESCRIPTION
**Pull request by robertwb**
These are enabled with the existing --profile_xxx flags.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by robertwb** [comment]
R: @tweise

**Comment by tweise** [comment]
@robertwb thanks! This looks good to me, although I have not tried to use it myself. I hope that @rakeshcusat and @akshaybalwally will do so and offer some feedback.

File: sdks/python/apache_beam/runners/worker/sdk_worker.py [hunk 1]
```diff
@@ -263,11 +269,23 @@ def process_bundle_progress(self, request, instruction_id):
     # It is an error to get progress for a not-in-flight bundle.
     processor = self.active_bundle_processors.get(request.instruction_reference)
     return beam_fn_api_pb2.InstructionResponse(
-        instruction_id=instruction_id,
+        instruction_id=abs(hash(instruction_id)),
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/worker/sdk_worker.py]: Why is this change needed?

File: sdks/python/apache_beam/utils/profiler.py [hunk 2]
```diff
@@ -59,14 +64,17 @@ def __exit__(self, *args):
 
     if self.profile_location and self.file_copy_fn:
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/profiler.py]: `self.file_copy_fn` can no longer be None and this check could be removed.

The behavior change here is that now it would not be possible to avoid this code path any more.

File: sdks/python/apache_beam/utils/profiler.py [hunk 3]
```diff
@@ -59,14 +64,17 @@ def __exit__(self, *args):
...
+          self.profile_location,
+          time.strftime(self.time_prefix + self.profile_id))
       fd, filename = tempfile.mkstemp()
-      self.profile.dump_stats(filename)
-      logging.info('Copying profiler data to: [%s]', dump_location)
-      self.file_copy_fn(filename, dump_location)  # pylint: disable=protected-access
-      os.close(fd)
-      os.remove(filename)
+      try:
+        os.close(fd)
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/profiler.py]: Would not this cause deletion of the temp file?

File: sdks/python/apache_beam/runners/portability/fn_api_runner.py [hunk 4]
```diff
@@ -235,11 +241,54 @@ def run_pipeline(self, pipeline):
     # are known to be KVs.
     from apache_beam.runners.dataflow.dataflow_runner import DataflowRunner
     pipeline.visit(DataflowRunner.group_by_key_input_visitor())
+    self._bundle_repeat = self._bundle_repeat or pipeline._options.view_as(
```
Comment (aaltay) [inline sdks/python/apache_beam/runners/portability/fn_api_runner.py]: Do runners have another way to access options? Do they need to use pipeline._options?

**Comment by robertwb** [comment]
Addressed comments and rebased. Though additional feedback would be good, I think this is at least a good start and we can continue to iterate with more experience. 

The tools to install are printed along with the profile if profiling is enabled and they are not available.

**Comment by tweise** [comment]
Would be nice to mention this in the docs as follow-up.

**Comment by robertwb** [comment]
Run Python PreCommit

**Comment by robertwb** [comment]
Running precommits again before merging just to be sure due to issues with  #6903.
